src/components/transactions/TxList/index.tsx

```diff
@@ -29,38 +32,59 @@ export const TxListGrid = ({ children }: { children: (ReactElement | null)[] }):
 
 const TxList = ({ items }: TxListProps): ReactElement => {
   const router = useRouter()
-  // Ensure list always starts with a date label
-  const list = useMemo(() => {
-    const firstDateLabelIndex = items.findIndex(isDateLabel)
-    const firstTxIndex = items.findIndex(isTransactionListItem)
-    const shouldPrependDateLabel =
-      (firstDateLabelIndex === -1 || firstDateLabelIndex > firstTxIndex) && firstTxIndex !== -1
 
-    if (!shouldPrependDateLabel) {
+  const list = useMemo(() => {
+    if (!hasTxFilterQuery(router.query)) {
       return items
     }
 
+    // Filtered transaction lists do not contain date labels
+    // Prepend initial date label to list
+    const firstTxIndex = items.findIndex(isTransactionListItem)
+
     const dateLabel: DateLabel = {
-      type: 'DATE_LABEL',
+      type: TransactionListItemType.DATE_LABEL,
       timestamp: (items[firstTxIndex] as Transaction).transaction.timestamp,
     }
+    const prependedItems = [dateLabel, ...items]
 
-    return [dateLabel, ...items]
-  }, [items])
+    // Insert date labels between transactions on different days
+    return prependedItems.reduce<TransactionListItem[]>((acc, current, i, arr) => {
+      const prev = acc[i - 1]
+      const isLastItem = i === arr.length - 1
+
+      if (
+        isLastItem ||
+        !prev ||
+        !isTransactionListItem(prev) ||
+        !isTransactionListItem(current) ||
+        // TODO: Make comparison in UTC
+        isSameDay(prev.transaction.timestamp, current.transaction.timestamp)
+      ) {
+        return acc.concat(current)
+      }
+
+      const dateLabel: DateLabel = {
+        type: TransactionListItemType.DATE_LABEL,
+        timestamp: current.transaction.timestamp,
+      }
+      return acc.concat(dateLabel)
+    }, [])
+  }, [items, router.query])
 
   const listWithGroupedItems: (TransactionListItem | Transaction[])[] = useMemo(() => {
-    return list.reduce((acc: (TransactionListItem | Transaction[])[], current) => {
+    return list.reduce((acc: (TransactionListItem | Transaction[])[], current, i) => {
       if (isConflictHeaderListItem(current)) {
-        return [...acc, []]
+        return acc.concat([[]])
```

Comment:
I've pivoted to using `concat` because it's faster with larger arrays.

Reply:
I understand, but why is it adding an array in the first place?

Reply:
It's seemingly for [conflicting transactions](https://github.com/safe-global/web-core/blob/main/src/components/transactions/TxList/index.tsx#L75-L77).

Reply:
Gotcha, thanks.